crates/bevy_ecs/src/schedule/stepping.rs

```diff
@@ -0,0 +1,1409 @@
+        };
+
+        // update the stepping frame cursor based on if there are any systems
+        // remaining to be run in the schedule
+        // Note: Don't try to detect the end of the render frame here using the
+        // schedule index.  We don't know all schedules have been added to the
+        // schedule_order, so only next_frame() knows its safe to reset the
+        // cursor.
+        match next_system {
+            Some(i) => self.cursor.system = i,
+            None => {
+                let index = cursor.schedule + 1;
+                self.cursor = Cursor {
+                    schedule: index,
+                    system: self.first_system_index_for_schedule(index),
+                };
+
+                #[cfg(test)]
+                debug!("advanced schedule index: {} -> {}", cursor.schedule, index);
+            }
+        }
+
+        Some(list)
+    }
+}
+
+#[derive(Default)]
+struct ScheduleState {
+    /// per-system [`SystemBehavior`]
+    behaviors: HashMap<NodeId, SystemBehavior>,
+
+    /// order of NodeIds in the schedule
+    ///
+    /// This is a cached copy of SystemExecutable.system_ids. We need it
+    /// available here to be accessed by Stepping::cursor() so we can return
+    /// NodeIds to the caller.
+    node_ids: Vec<NodeId>,
+
+    /// changes to system behavior that should be applied the next time
+    /// [`ScheduleState::skipped_systems()`] is called
+    behavior_updates: HashMap<TypeId, Option<SystemBehavior>>,
+
+    /// This field contains the first steppable system in the schedule.
+    first: Option<usize>,
+}
+
+impl ScheduleState {
+    // set the stepping behavior for a system in this schedule
+    fn set_behavior(&mut self, system: SystemIdentifier, behavior: SystemBehavior) {
+        self.first = None;
+        match system {
+            SystemIdentifier::Node(node_id) => {
+                self.behaviors.insert(node_id, behavior);
+            }
+            // Behaviors are indexed by NodeId, but we cannot map a system
+            // TypeId to a NodeId without the `Schedule`.  So queue this update
+            // to be processed the next time `skipped_systems()` is called.
+            SystemIdentifier::Type(type_id) => {
+                self.behavior_updates.insert(type_id, Some(behavior));
+            }
+        }
+    }
+
+    // clear the stepping behavior for a system in this schedule
+    fn clear_behavior(&mut self, system: SystemIdentifier) {
+        self.first = None;
+        match system {
+            SystemIdentifier::Node(node_id) => {
+                self.behaviors.remove(&node_id);
+            }
+            // queue TypeId updates to be processed later when we have Schedule
+            SystemIdentifier::Type(type_id) => {
+                self.behavior_updates.insert(type_id, None);
+            }
+        }
+    }
+
+    // clear all system behaviors
+    fn clear_behaviors(&mut self) {
+        self.behaviors.clear();
```

Comment:
why arent `behavior_updates` cleared here too?

Reply:
Great catch!  Queue Schedule changes were being applied after the clear.

I've fixed it and added a test case to cover it.  Added an additional test case to also verify the reverse order (clear_schedule(), never_run()) executes correctly.  Also confirmed the other queuing behavior (`Stepping.updates`) does not have a similar drain functionality.